notebooks/notebook_template_review.py

```diff
@@ -154,268 +154,324 @@ class ErrorCode(Enum):
     """
     Abstract class for defining notebook conformance rules
     """
-    cell_index = 0
-    title = ''
-    git_link = None
-    colab_link = None
-    workbench_link = None
-    desc = ''
-    uses = ''
-    steps = ''
-    costs = []
-    
-    @staticmethod
-    def init():
-        NotebookRule.cell_index = 0
-        NotebookRule.title = ''
-        NotebookRule.git_link = None
-        NotebookRule.colab_link = None
-        NotebookRule.workbench_link = None
-        NotebookRule.desc = ''
-        NotebookRule.uses = ''
-        NotebookRule.steps = ''
-        NotebookRule.costs = ''
-    
     @abstractmethod
-    def validate(self, path: str, cells: list) -> None:
+    def validate(self, notebook: Notebook):
+        '''
+        Applies cell specific rules to validate whether the cell 
+        does or does not conform to the rules.
+        '''
         pass
 
 
+
 class CopyrightRule(NotebookRule):
-    def validate(self, path: str, cells: list) -> None:
+    def validate(self, notebook: Notebook):
         """
         Parse the copyright cell
-
-        path: used only for reporting an error
-        cells: The content cells (JSON) for the notebook
         """
-        cell, cell_index = get_cell(path, cells, NotebookRule.cell_index)
+        cell = notebook.get()
         if not 'Copyright' in cell['source'][0]:
-            report_error(path, ErrorCode.ERROR_COPYRIGHT, "missing copyright cell")
-            
-        NotebookRule.cell_index = cell_index
+            self.report_error(ErrorCode.ERROR_COPYRIGHT, "missing copyright cell")
 
 
 class NoticesRule(NotebookRule):
-    def validate(self, path: str, cells: list) -> None:
+    def validate(self, notebook: Notebook):
         """
         Parse the (optional) notices cell
-
-        path: used only for reporting an error
-        cells: The content cells (JSON) for the notebook
         """
-        cell, cell_index = get_cell(path, cells, NotebookRule.cell_index)
+        cell = notebook.peek()
         if cell['source'][0].startswith('This notebook'):
-             NotebookRule.cell_index = cell_index
-        else:
-            NotebookRule.cell_index = cell_index - 1
+            notebook.pop()
 
 
-class TitleRule(NotebookRule):
-    def validate(self, path: str, cells: list) -> None: 
+class TitleRule(NotebookRule): 
+    def validate(self, notebook: Notebook): 
         """
         Parse the title in the links cell
-
-        path: used only for reporting an error
-        cells: The content cells (JSON) for the notebook
         """
-        cell, cell_index = get_cell(path, cells, NotebookRule.cell_index)
+        cell = notebook.peek()
         if not cell['source'][0].startswith('# '):
-            report_error(path, ErrorCode.ERROR_TITLE_HEADING, "title cell must start with H1 heading")
-            NotebookRule.title = ''
+            notebook.report_error(ErrorCode.ERROR_TITLE_HEADING, "title cell must start with H1 heading")
```

Comment:
Instead of storing state in a notebook object, it's much simpler to just loop over notebook cells and return results.
The less places you store state (especially, mutable state like this), the better.